src/mongoose_batch_worker.erl

```diff
@@ -0,0 +1,136 @@
+-module(mongoose_batch_worker).
+
+-behaviour(gen_server).
+
+-include("mongoose_logger.hrl").
+
+%% gen_server callbacks
+-export([init/1,
+         handle_call/3,
+         handle_cast/2,
+         handle_info/2,
+         terminate/2,
+         code_change/3,
+         format_status/2]).
+
+-record(state, {
+          host_type :: mongooseim:host_type(),
+          batch_size :: non_neg_integer(),
+          flush_interval :: non_neg_integer(), %% milliseconds
+          flush_interval_tref :: undefined | reference(),
+          flush_callback = fun(_, _) -> ok end :: fun((list(), map()) -> ok | {error, term()}),
+          flush_queue = [] :: list(),
+          flush_extra = #{} :: map()
+         }).
+-type state() :: #state{}.
+
+%% gen_server callbacks
+-spec init({mongooseim:host_type(),
+            pos_integer(),
+            pos_integer(),
+            fun((list(), map()) -> ok | {error, term()}),
+            map()}) -> {ok, state()}.
+init({HostType, Interval, MaxSize, FlushCallback, FlushExtra}) ->
+    ?LOG_INFO(#{what => starting_worker, host_type => HostType, pool_id => FlushCallback}),
+    {ok, #state{host_type = HostType,
+                batch_size = MaxSize,
+                flush_interval = Interval,
+                flush_callback = FlushCallback,
+                flush_extra = FlushExtra}}.
+
+-spec handle_call(term(), {pid(), term()}, state()) -> {reply, term(), state()}.
+handle_call(Msg, From, State) ->
+    ?UNEXPECTED_CALL(Msg, From),
+    {reply, unexpected_call, State}.
+
+-spec handle_cast(task | term(), state()) -> {noreply, state()}.
+handle_cast({task, Task}, State) ->
+    {noreply, handle_task(Task, State)};
+handle_cast(Msg, State) ->
+    ?UNEXPECTED_CAST(Msg),
+    {noreply, State}.
+
+-spec handle_info(flush | term(), state()) -> {noreply, state()}.
+handle_info(flush, State) ->
+    {noreply, run_flush(State)};
+handle_info({garbage_collect, asynchronous_gc_triggered, true}, State) ->
+    {noreply, State};
+handle_info(Msg, State) ->
+    ?UNEXPECTED_INFO(Msg),
+    {noreply, State}.
+
+-spec terminate(term(), state()) -> term().
+terminate(_Reason, State) ->
+    ?LOG_INFO(log_fields(State, #{what => stopping_worker})),
+    case [] =:= State#state.flush_queue of
+        true -> ok;
+        false ->
+            do_run_flush(State),
+            ?LOG_WARNING(log_fields(State, #{what => worker_terminate_requires_flush}))
+    end.
+
+-spec code_change(term(), state(), term()) -> {ok, state()}.
+code_change(_OldVsn, State, _Extra) ->
+    {ok, State}.
+
+% Don't leak the tasks to logs, can contain private information
+format_status(_Opt, [_PDict, State | _]) ->
+    [{data, [{"State", State#state{flush_queue = []}},
```

Comment:
`flush_queue = censored` ;) 
actually, setting it to a valid value here (i.e. []) could be pretty confusing, when people try to call for status.

could be `= hidden or sanitized`

Reply:
censored returns a dialyzer error here, but can add a type for that in the typedef

Reply:
sad :(

Actually, just actually, why not to convert `State` into a map in format status? :) 
It is much easier to read ;)

Reply:
I think a super canonical example for format_status was actually converting state into a proplist :) 
(it is what I think about when see that gen-server callback).